web/src/compose_ui.js

```diff
@@ -550,6 +550,138 @@ export function format_text($textarea, type, inserted_content) {
         wrapSelection(field, syntax_start, syntax_end);
     };
 
+    const format_spoiler = () => {
+        let spoiler_syntax_start = "```spoiler \n";
+        const spoiler_syntax_start_without_break = "```spoiler ";
+        let spoiler_syntax_end = "\n```";
+
+        // Captures:
+        // ```spoiler |
+        // <text>
+        // ```
+        // (with | as placeholder for the cursor)
+        const is_cursor_at_header_position = () =>
+            range.length === 0 &&
+            text.slice(range.start - spoiler_syntax_start_without_break.length, range.start) ===
+                spoiler_syntax_start_without_break;
+
+        // Captures:
+        // ```spoiler <header>
+        // <text>
+        // ```
+        // with <text> selected
+        const is_content_selected = () =>
+            text.slice(range.end, range.end + spoiler_syntax_end.length) === spoiler_syntax_end &&
+            text[range.start - 1] === "\n" &&
```

Comment:
Does this need a `range.start > 0` check?

Reply:
Added this and a couple more range checks for spoilers and links